Add blog post for v6.0.0-preview (#107)
EventStore/eventstore.org: _posts/2019-12-16-event-store-6.0.0-preview-release.md

With this release we want to show you where we are planning on taking Event Store in the future.

This release is not intended to be used in production and is still rough around the edges, however we wanted to provide you with the release bits and we welcome your feedback.

If you encounter any issues, please don’t hesitate to open an issue on GitHub if there isn’t one already.

## Important!

With the preview release of Event Store, Event Store will only expose the external HTTP interface over HTTPS.

This requires an TLS certificate, but for ease of use we have introduced a development mode which will use a self signed certificate intended for development use only.

# What’s new in Event Store

There are a number of exciting features included in this preview:
- The move to .NET Core 3.1
- A gRPC client
- Server Side Filtering
- Read-Only Replicas
- Master Resignation
- Restarting the Projection Subsystem
- Pinned by Correlation Id strategy for Competing Consumers

Given that this is a preview release, there is no documentation on the new features at the moment.

We have included details on how to get started with each feature further along in this blog post.  

> Additionally, you can reach out to us over github or google groups to ask questions or report any issues.

Reviewer: ```suggestion
Additionally, you can reach out to us over [github](https://github.com/EventStore/EventStore) or [google groups](https://groups.google.com/forum/#!forum/event-store) to ask questions or report any issues.
```

Reviewer 2: I would rather people approached us via github issues only really. otherwise we split conversations...

## .NET Core 3.1

> Event Store 6 runs only on Core CLR 3.1. Support for both .NET Framework and Mono have been dropped.  

Reviewer: ```suggestion
Event Store 6 only runs on Core CLR 3.1. Support for both .NET Framework and Mono have been dropped.  
```

---
title: "Event Store 6.0.0 Preview 1"
author: "Hayley Campbell"
layout: blog-post
---

> We are very excited to announce the first preview release of Event Store 6.0.0!  

Reviewer 3: ```suggestion
We are excited to announce the first preview release of Event Store 6.0.0!  
```

With this release we want to show you where we are planning on taking Event Store in the future.

> This release is not intended to be used in production and is still rough around the edges, however we wanted to provide you with the release bits and we welcome your feedback.

Reviewer 3: ```suggestion
This release is not intended to be used in production and is still rough around the edges, however we want to provide you with a release candidate, and we welcome your feedback.
```

> If you encounter any issues, please don’t hesitate to open an issue on GitHub if there isn’t one already.

Reviewer 3: ```suggestion
If you encounter any issues, please don’t hesitate to [open an issue](https://github.com/EventStore/EventStore/issues/new) on GitHub if there isn’t one already.
```

## Important!

With the preview release of Event Store, Event Store will only expose the external HTTP interface over HTTPS.

> This requires a TLS certificate, but for ease of use we have introduced a development mode which will use a self signed certificate intended for development use only. Development mode can be enabled by specifying the `--dev` option when starting Event Store.

Reviewer 3: ```suggestion
This requires a TLS certificate, but for ease of use we have introduced a development mode which uses a self signed certificate intended for development use only. Development mode can be enabled by specifying the `--dev` option when starting Event Store.
```

> # What’s new in Event Store

Reviewer 3: ```suggestion
## What’s new in Event Store
```

There are a number of exciting features included in this preview:
- The move to .NET Core 3.1
- A gRPC client
- Server Side Filtering
- Read-Only Replicas
- Master Resignation
- Restarting the Projection Subsystem
- Pinned by Correlation Id strategy for Competing Consumers

> Given that this is a preview release, there is no documentation on the new features at the moment.

Reviewer 2: Maybe it would be better to say "We are working on the documenting the new features fully for release. For the preview, a we have included an overview and brief guide to each feature in this post"

We have included details on how to get started with each feature further along in this blog post.  
Additionally, you can reach out to us over github or google groups to ask questions or report any issues.

> ## .NET Core 3.1

Reviewer 3: ```suggestion
### .NET Core 3.1
```

Event Store 6 runs only on Core CLR 3.1.  
Support for both .NET Framework and Mono have been dropped.  

> This has provided a variety of benefits, and is one of the most requested changes we have ever made.

Reviewer 2: This is one of the most requested changes we have ever had and we are pleased to be starting the process of finally  delivering it with this release.

> Much of the platform-specific code has been removed, and we benefit from the focus on performance that Microsoft has given the Core CLR.   We now use the Kestrel HTTP server, which has significant benefits over HttpListener which was used up to version 5.

Reviewer 3: ```suggestion
Much of the platform-specific code has been removed, and we benefit from the focus on performance that Microsoft has given the Core CLR. We now use the Kestrel HTTP server, which has significant benefits over HttpListener which was used up to version 5.
```

## gRPC Client API

Since version 1.0 of Event Store, two client protocols have been supported - Atompub over HTTP, and full-duplex protocol-buffers-over-TCP.

Each of these has advantages - the Protocol Buffers API is better for latency-sensitive workloads, while the Atom API gives more flexibility for cache design. However, both require a lot of work on each platform to implement.

To improve this situation, we have added a new default client protocol in Event Store 6, using gRPC - a widely adopted standard which is interoperable across lots of platforms with minimal effort.

In Preview 1 of Event Store 6, we are shipping a .NET SDK for the new gRPC protocol. With the final version of Event Store 6, we will also ship first party, commercially supported client SDKs for Java, Node.js and Go in addition to .NET.

The Event Store .NET gRPC SDK is available on [NuGet](https://www.nuget.org/packages/EventStore.Grpc.Client/).

The following code snippets show how to perform some common operations with the .NET SDK:

### Create the Client

> ```c#

Reviewer 3: ```suggestion
```cpp
```

var eventStoreClient = new EventStore.Grpc.EventStoreGrpcClient(new Uri("https://localhost:2113/"));
```

**Note:**  
In order for the gRPC client to connect successfully, the TLS certificate used by Event Store must be trusted by your machine.  
When running in development mode, this would be the development certificates found in `dev-ca` in the extracted directory.

Alternatively, you can disable the validation checks by specifying a custom HttpClient when creating the client, for example:

> ```c#

Reviewer 3: ```suggestion
```cpp
```

var eventStoreClient = new EventStore.Grpc.EventStoreGrpcClient(new Uri("https://localhost:2113/"),
    () => new HttpClient(new SocketsHttpHandler
    {
        SslOptions = new SslClientAuthenticationOptions
        {
            RemoteCertificateValidationCallback = delegate { return true; }
        }
    }));
```

### Writing to a Stream with Expected Version Any

> ```c#

Reviewer 3: ```suggestion
```cpp
```

Reviewer 3: etc… I'm not sure if `c#` is supported, I have never seen it used like that.

var eventData = new EventData(Uuid.NewUuid(), "type", JsonSerializer.SerializeToUtf8Bytes(new SomeEvent()));
await eventStoreClient.AppendToStreamAsync(
  "stream", AnyStreamRevision.Any, new[] { eventData }, cancellationToken: cancellationToken);
```

### Writing to a Stream with Expected Version

```c#

> var eventData = new EventData(Uuid.NewUuid(), "type", JsonSerializer.SerializeToUtf8Bytes(new SomeEvent()));

Reviewer 4: One suggestion, I think it's better to use  the full type name here i.e. `System.Text.Json.JsonSerializer.SerializeToUtf8Bytes`, otherwise there might be confusion as `Newtonsoft.Json` has the same name here.

The intention here is to remove the need for having to setup and use projections when you only want to subscribe to a stream containing specific events.

New methods and overloads have been added to the .NET client API for reads and subscriptions.  
New HTTP endpoints have been added as well for reading only.

The API differs depending on whether you use the TCP or gRPC client. Below are some examples to help you get started.

### TCP Client

The server side filter is created using the new Filter class and you can choose to match either the event type or stream name against a prefix or regular expression.

```c#
var filter = Filter.EventType.Prefix("SimpleEvent");
```

That filter can then be used when making calls to either `FilteredReadAllEventsForwardAsync` or `FilteredReadAllEventsBackwardAsync` on the connection.

```c#
var readEvents = await connection.ReadAllEventsForwardFilteredAsync(
    position: Position.Start, 
    maxCount: 4096, 
    resolveLinkTos: false,
    eventFilter: eventFilter,
    maxSearchWindow: 1000);
```

There is also an additional overload where you can provide a max search window argument.  

> This specifies the maximum number of events the server will search through before returning the slice.

Reviewer 2: If no events are found within the window, an empty slice will be returned... @mat-mcloughlin ?

To address both of these concerns, we have introduced the ability to mark a node as a read-only replica which avoids these scenarios. This type of node will not partake in elections, and will not be promotable to a clone, slave or master node.

To start a node as a read-only replica, the `ReadOnlyReplica` option has been introduced.

```bash
%> EventStore.ClusterNode.exe --read-only-replica
```

The Event Store Client has also been adjusted to allow the user to set the node preference to ReadOnlyReplica.

```c#
var settings = EventStore.ClientAPI.ConnectionSettings.Create()
.PerformOnAnyNode()
.PreferReadOnlyReplica();
```

In the above example, you will notice that we also specify `PerformOnAnyNode`.  
This is required as by default operations want to be performed on the master node only, so if this option is not set, the client will reconnect to the master node even if `PreferReadOnlyReplica` is specified.

Once a node is connected to a ReadOnlyReplica node with the `PreferReadOnlyReplica` option set, a client will not reconnect to any other type of node.

> While adding clones have not yet been deprecated, it will be in a future release.

Reviewer 2: has not yet
